ebos/ecloutputblackoilmodule.hh

```diff
@@ -1077,12 +1078,50 @@ private:
                 if (!this->fip_[Inplace::Phase::OilInGasPhase].empty())
                     this->fip_[Inplace::Phase::OilInGasPhase][globalDofIdx] = oilInPlaceGas;
 
+                // CO2STORE OIL + GAS
+                if (!this->fip_[Inplace::Phase::CO2InWaterPhase].empty())
+                    this->fip_[Inplace::Phase::CO2InWaterPhase][globalDofIdx] = gasInPlaceLiquid;
+
                 // Add dissolved gas and vaporized oil to total Fip
                 if (!this->fip_[Inplace::Phase::OIL].empty())
                     this->fip_[Inplace::Phase::OIL][globalDofIdx] += oilInPlaceGas;
                 if (!this->fip_[Inplace::Phase::GAS].empty())
                     this->fip_[Inplace::Phase::GAS][globalDofIdx] += gasInPlaceLiquid;
             }
+
+            if (FluidSystem::phaseIsActive(waterPhaseIdx) && FluidSystem::phaseIsActive(gasPhaseIdx)) {
+                // Gas dissolved in water and vaporized water
+                Scalar gasInPlaceWater = getValue(fs.Rsw()) * fip[waterPhaseIdx];
+                Scalar waterInPlaceGas = getValue(fs.Rvw()) * fip[gasPhaseIdx];
+
+                if (!this->fip_[Inplace::Phase::CO2InWaterPhase].empty())
+                    this->fip_[Inplace::Phase::CO2InWaterPhase][globalDofIdx] = gasInPlaceWater;
+
+                if (!this->fip_[Inplace::Phase::WaterInGasPhase].empty())
+                    this->fip_[Inplace::Phase::WaterInGasPhase][globalDofIdx] = waterInPlaceGas;
+
+                if (!this->fip_[Inplace::Phase::WaterInWaterPhase].empty())
+                    this->fip_[Inplace::Phase::WaterInWaterPhase][globalDofIdx] = fip[waterPhaseIdx];
+
+                // Add dissolved gas and vaporized water to total Fip
+                if (!this->fip_[Inplace::Phase::WATER].empty())
+                    this->fip_[Inplace::Phase::WATER][globalDofIdx] += waterInPlaceGas;
+                if (!this->fip_[Inplace::Phase::GAS].empty())
+                    this->fip_[Inplace::Phase::GAS][globalDofIdx] += gasInPlaceWater;
+            }
+
+            Scalar criticalGas = 0.1; //TODO get this from the tables
```

Comment:
@bska What is the best way to get the critical gas saturation here? Make it a member in this class and initialize from eclState via tables and satprop code?

Reply:
> What is the best way to get the critical gas saturation here

Just to be sure: By "critical gas saturation" you mean the saturation at which the gas phase becomes mobile&mdash;i.e., exceeds the critical relative permeability tolerance ("TOLCRIT")?  If so, you typically need to care about end-point scaling and/or hysteresis which varies by cell and/or system state.  In that case, I think the most general approach is to go through
```
simulator_.problem().materialLawManager()
```
and effectively do the same that we do when computing the relative permeability.  Storing this as a data member will just duplicate what we already have.